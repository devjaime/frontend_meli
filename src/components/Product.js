import React, { useEffect, useState } from 'react';
import useGetProduct from '../api/useGetProduct';
import { formatCurrency } from '../utils/misc/formatter';
import loader from '../assets/loader.gif';
import iconShipping from '../assets/images/ic_shipping.png';

export const Product = (props) => {
  const itemId = props.match.params.id;
  const propsState = props.location.state;
  const categories = propsState && propsState.categories;
  const itemCategory = propsState && propsState.itemCategory;

  const [product, setProduct] = useState({});

  const getItem = useGetProduct(itemId);
  useEffect(() => {
    if (getItem.response) setProduct(getItem.response);
  }, [getItem.response]);

  return (
    <div>
      {product && product.item
        ? (
          <div>
            <div className="item-categories-container">
              {itemCategory || categories || ''}
            </div>
            <div className="product-container">
              <div className="product-card">
                <div>
                  <img src={product.picture} alt={product.name} />
                </div>
                <div className="product-info">
                  <div className="product-state">
                    {product.condition === 'new' ? 'Nuevo' : 'Usado'}
                    {' - '}
                    {`${product.sold_quantity} vendidos`}
                  </div>
                  <div className="product-name">
                    {product.item.title}
                  </div>
                  <div className="product-price">
                    {`$${formatCurrency(product.item.price.amount, product.item.price.decimals)}`}
                    {product.free_shipping
                                        && <img src={iconShipping} alt="free_shipping" title="Envío gratis" />}
                  </div>
                  <div className="buy-button-container">
                    <button type="button">Comprar</button>
                  </div>
                </div>
              </div>
              <div className="product-description">
                <div className="title">Descripción del producto</div>
                <div className="description">{product.description}</div>
              </div>
            </div>
          </div>
        )
        : (
          <div className="loader">
            <img src={loader} alt="loader" />
          </div>
        )}
    </div>
  );
};

export default Product;
