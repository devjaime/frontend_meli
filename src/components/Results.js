import React, { useState, useEffect } from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import useSearchItems from '../../src/api/useSearchItems';
import iconShipping from '../assets/images/ic_shipping.png';
import { formatCurrency } from '../utils/misc/formatter';
import loader from '../assets/loader.gif';

function useQuery() {
    return new URLSearchParams(useLocation().search);
}

export const Results = () => {
    const [results, setResults] = useState([]);
    const history = useHistory();
    let query = useQuery();
    let searchQuery = query.get("search");
    const searchItems = useSearchItems(searchQuery);
    
    useEffect(() => {
        if (searchItems.response) {
            setResults(searchItems.response);
        }
    }, [searchItems.response]);

    const renderItem = (itemId, itemCategory) => {
        history.push({
            pathname: `/items/${itemId}`,
            state: {
                categories: results.categories.join(' > '), 
                itemCategory: itemCategory
            }
        });
    };

    return (
        <div>
            {results && results.items && results.items.length
                ?
                <div className='results-container'>
                    {results && results.categories &&
                        <div className='categories-container'>
                            {results && results.categories &&
                                results.categories.join(' > ')
                            }
                        </div>
                    }
                    {results && results.items && results.items.map(item => (
                        <div key={item.id} className='item-card' onClick={() => renderItem(item.id, item.category)}>
                            <div className='left-side-card'>
                                <img src={item.picture} alt={item.title} />
                            </div>
                            <div className='right-side-card'>
                                <div className='price-container'>
                                    {`$ ${formatCurrency(item.price.amount, item.price.decimals)}`}
                                </div>
                                {item.free_shipping &&
                                    <img src={iconShipping} alt="free_shipping" title="EnvÃ­o gratis" />
                                }
                                <div className='location'>{item.location}</div>
                                <div className='title-container'>
                                    {item.title}
                                </div>
                            </div>
                        </div>
                    ))}
                </div>
                :
                <div className='loader'>
                    <img src={loader} alt="loader"/>
                </div>
            }

        </div>
    );
}

export default Results;