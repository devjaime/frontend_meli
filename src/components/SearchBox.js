import React, { useState, useRef, useEffect } from 'react';
import { useLocation, useHistory } from 'react-router-dom';

import logoML from '../assets/images/logo_ml.png';
import iconSearch from '../assets/images/ic_search.png';

function useQuery() {
  return new URLSearchParams(useLocation().search);
}
export const SearchBox = () => {
  const firstRender = useRef(true);
  const history = useHistory();

  const [item, setItem] = useState('');

  const query = useQuery();
  const searchQuery = query.get('search');

  useEffect(() => {
    if (firstRender.current) firstRender.current = false;
    if (searchQuery) setItem(searchQuery);
  }, [searchQuery]);

  const handleChange = (e) => {
    setItem(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (item) {
      history.push({
        pathname: '/items',
        search: `?search=${item}`,
      });
    }
  };

  const goToHome = () => {
    // eslint-disable-next-line no-alert
    alert('boton busqueda');
    history.push('/');
  };

  return (
    <header className="header">
      <div>
        <div
          tabIndex={0}
          onClick={() => goToHome()}
          onKeyDown={() => goToHome()}
          role="button"
        >
          <img
            src={logoML}
            alt="logo_ml"
            title="Mercado Libre - Donde comprar y vender de todo"
            className="meli-logo"
          />
        </div>
        <form onSubmit={(e) => handleSubmit(e)}>
          <input
            type="text"
            id="serach_input"
            name="search"
            value={item}
            placeholder="Buscar productos, marcas y mÃ¡s"
            onChange={(e) => handleChange(e)}
          />
          <button type="submit">
            <img src={iconSearch} alt="ic_search" title="Buscar" />
          </button>
        </form>
      </div>
    </header>
  );
};

export default SearchBox;
