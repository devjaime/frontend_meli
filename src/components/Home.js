import React from 'react';
import SearchBox from '../components/SearchBox';

function Home() {
    return (
        <SearchBox />
    );
}

export default Home;