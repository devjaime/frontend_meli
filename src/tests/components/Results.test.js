import React from 'react';
import '@testing-library/jest-dom';
import { mount } from "enzyme";
import { Results } from '../../components/Results';
import { MemoryRouter, Route } from "react-router-dom";
     

describe('Test in <Results/>', () => {

    test('should display the component correctly', () => {
        let testHistory;
        let testLocation;
        let History={"length":2,"action":"PUSH","location":{"pathname":"/items","search":"?search=monitor curvo","hash":"","key":"q06uc0"}};
        let Location={"pathname":"/items","search":"?search=monitor curvo","hash":"","key":"q06uc0"};
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/`]}>
            <Results />
            <Route
                path={`*`}
                render={routeProps => {
                testHistory = History
                testLocation = Location
                return null
                }}/>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(testHistory).toMatchSnapshot();
        expect(testLocation).toMatchSnapshot();
    })
})