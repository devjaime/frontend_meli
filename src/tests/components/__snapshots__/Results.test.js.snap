// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Results/> should display the component correctly 1`] = `
Array [
  <div>
    <div
      className="loader"
    >
      <img
        alt="loader"
        src="loader.gif"
      />
    </div>
  </div>,
  "",
]
`;

exports[`Test in <Results/> should display the component correctly 2`] = `
Object {
  "action": "PUSH",
  "length": 2,
  "location": Object {
    "hash": "",
    "key": "q06uc0",
    "pathname": "/items",
    "search": "?search=monitor curvo",
  },
}
`;

exports[`Test in <Results/> should display the component correctly 3`] = `
Object {
  "hash": "",
  "key": "q06uc0",
  "pathname": "/items",
  "search": "?search=monitor curvo",
}
`;
